cpp/src/arrow/array/data.cc

```diff
@@ -381,24 +381,29 @@ void ArraySpan::FillFromScalar(const Scalar& value) {
       this->child_data[i].FillFromScalar(*scalar.value[i]);
     }
   } else if (is_union(type_id)) {
+    // Dense union needs scratch space to store both offsets and a type code
+    struct UnionScratchSpace {
+      uint8_t type_code;
+      alignas(int32_t) uint8_t offsets[sizeof(int32_t) * 2];
```

Comment:
If, out of caution, we want these pointers to have 64-bit alignment like provided by `malloc`, the `type_code` should align to `int64_t`. As it is now, you're using the second and third 4-byte block of the scratch space, when you could use the third and fourth.

Reply:
```suggestion
      uint8_t type_code;
      alignas(int64_t) uint8_t offsets[sizeof(int32_t) * 2];
```

Reply:
That also does the trick :)